### ISSUE TITLE
Port exhaustion on Windows node makes DNS stop working on Windows Pods

### ISSUE THREAD

**Issue by asbjorne** [Azure/AKS#1375]
**What happened**:
DNS/name resolving (both internally and externally) stopped working on windows pods on a specific node. 

From pod (the error response is immediate)
```
PS C:\Program Files\Genus\Server> Resolve-DnsName -name www.google.com
Resolve-DnsName : www.google.com : This operation returned because the timeout period expired
At line:1 char:1
+ Resolve-DnsName -name www.google.com
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationTimeout: (www.google.com:String) [Resolve-DnsName], Win32Exception   
    + FullyQualifiedErrorId : ERROR_TIMEOUT,Microsoft.DnsClient.Commands.ResolveDnsName

PS C:\Program Files\Genus\Server> Resolve-DnsName kubernetes.default 
Resolve-DnsName : kubernetes.default : This operation returned because the timeout period expired 
At line:1 char:1
+ Resolve-DnsName kubernetes.default
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationTimeout: (kubernetes.default:String) [Resolve-DnsName], Win32Except  
   ion
    + FullyQualifiedErrorId : ERROR_TIMEOUT,Microsoft.DnsClient.Commands.ResolveDnsName
```

DNS works properly when running nslookup from the node.
```
PS C:\k> Resolve-DnsName www.google.com

Name                                           Type   TTL   Section    IPAddress
----                                           ----   ---   -------    ---------
www.google.com                                 AAAA   142   Answer     2a00:1450:400e:808::2004
www.google.com                                 A      142   Answer     172.217.19.196
```
Looking into kube-proxy logs I found:
`E0107 21:00:30.725221    6840 proxier.go:1212] Policy creation failed: hcnCreateLoadBalancer failed in Win32: The specified port already exists. (0x803b0013) {"Success":false,"Error":"The specified port already exists. ","ErrorCode":2151350291}`

I then ran CollectLogs.ps1 on the node and found this in reservedports.txt (attached):
```
ERROR: Running out of ephemeral ports. The ephemeral ports range doesn't have enough resources to allow allocating 64 contiguous TCP ports.
Couldn't reserve more than 0 ranges of 64 ports

```

This correlates to the known issue:
https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/common-problems#load-balancers-are-plumbed-inconsistently-across-the-cluster-nodes

I then checked which ports are in use with netstat (see attached netstat.txt).
This shows about 30 ports in the user range (49152-65535).

Looking at excludedports.txt (attached) I find that almost all ports are reserved which could be the cause of the exhaustion.

Any ideas to what could cause so many ports to be reserved?

We've experienced this several times and also been in contact with support on this issue (119111922001016). Re-imaging the node makes it work again, but I want to submit this issue so we can maybe solve the actual problem.

I have all logs from CollectLogs.ps1 if needed.

[reservedports.txt](https://github.com/Azure/AKS/files/4034736/reservedports.txt)
[dynamicportrange.txt](https://github.com/Azure/AKS/files/4034737/dynamicportrange.txt)
[excludedportrange.txt](https://github.com/Azure/AKS/files/4034738/excludedportrange.txt)
[netstat-run-manually.txt](https://github.com/Azure/AKS/files/4034739/netstat-run-manually.txt)
[ports.txt](https://github.com/Azure/AKS/files/4034740/ports.txt)

**What you expected to happen**:
DNS/name resolving working in the pods as usual.

**How to reproduce it**:
The node is still running in AKS and can be accessed for more debugging and logging.

Not sure how to reproduce the port exhaustion, but it has occured at least 5 times on different windows nodes. This time it occured just a few hours after creation of the node. Earlier the node has been running without problems for 7-12 days before the problem occurs. 

**Environment**:
- Kubernetes version: 1.17.0 (preview) (also experienced issue with 1.14.x and 1.15.5)
- Size of cluster: 10 Linux (Standard_DS2_v2 and Standard_DS3_v2) and 2 Windows worker nodes (Standard_B8ms)
- General description of workloads in the cluster: Node.js, Nginx, .Net Core (Linux) and .Net Framework (Windows) pods

**Comment by zhiweiv** [comment]
I got same problem several times too, my SR is 119112223001068, it is still pending at azure support side.

**Comment by PashaPash** [comment]
Got the same issue. For me it manifested in infinite pod restarts with the following message:

> pod sandbox changed, it will be killed and re-created.

It was quite easy to reproduce - on a 10lin+2win node cluster with a 100+ k8s services scaling any any windows deployment to 20+ pods hits that dynamic port limit, causing pod restarts.

Manually connecting to each windows node and expanding dynamic port range helps, at least until the next re-image/update. You can use VMSS Boot diag or SSH to get to the node console, then run

    netsh int ipv4 set dynamicportrange TCP start=1025 num=64511

That will max out dynamic port range.

SSH Setup for windows node:

> Download a Microsoft published yaml file we will use:
> 
>     wget https://raw.githubusercontent.com/Microsoft/SDN/master/Kubernetes/flannel/l2bridge/manifests/simpleweb.yml -O win-webserver.yaml
> 
> Now apply the yaml to deploy the pod:
> 
>     kubectl apply -f win-webserver.yaml
> 
> You can run this command to wait for the pod to spin up:
> 
>     watch kubectl get pods -o wide
> 
> SSH Into a Node
> 
>     kubectl get nodes -o wide
> 
> Now get the name of the pod just deployed, hit: it will start with win-webserver
> 
>     kubectl get pod -o=name | grep win
> 
> Lets exec to that pod and ssh directly into the Node we gathered above.
> Note you will need to substitute the IP address for your node you get with the command
> 
>     kubectl exec -it win-webserver-7812312-211 -- ssh 10.240.0.5 -l azureuser
> 

It will ask you for azureuser password. If you don't have it, then using VMSS Password Reset combined with enabling VMSS Boot diag will be less time consuming than SSH  - as you will have to reimage nodes to apply password reset anyway.

**Comment by zhiweiv** [comment]
As described in https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/common-problems#load-balancers-are-plumbed-inconsistently-across-the-cluster-nodes: 

> For a permanent solution, kube-proxy load balancing should be set to DSR mode.

I guess DSR will probably be available in Windows Server 2022, so we have to wait 2 years for k8s Windows to production ready?

**Comment by MikkelHegn** [comment]
Following up internally as I know we've looked in to increasing the range in non-DSR mode for certain scenarios. Will post back once I get more details.

**Comment by daschott** [comment]
DSR mode is already available in 1903. But some traffic paths will not work. The following should work on 1903/1909 DSR mode:
o	container to container traffic
o	container to service traffic
o	container to external traffic (Internet)

The following will not work on DSR mode K8s services until support is added in the next SAC release:
o	Host -> service traffic
o	Host -> pod traffic
o	Host -> NodePort traffic
o	External client -> NodePort traffic

There are also plans to backport DSR mode feature all the way back to 1809, but this will not happen earlier than mid-2020.

**Comment by zhiweiv** [comment]
@daschott 
> The following will not work on DSR mode K8s services until support is added in the next SAC release:
> o Host -> service traffic
> o Host -> pod traffic
> o Host -> NodePort traffic
> o External client -> NodePort traffic

These paths are absolutely necessary. Just to confirm, they will work on Windows Server version 2003?

**Comment by asbjorne** [comment]
@flabbergasted I see your mentioned documentation issue. 
Does the 250 service per cluster limitation refer to kubernetes services or pods? 
And if we stay under the limit we should be better off then now? 

We're currently running 256 pods in our clusters where we experience port exhaustion.
As a temporary fix we've increased the port range as advised here on the the thread.

**Comment by MikkelHegn** [comment]
It's the number of Kubernetes services across the cluster, which Windows need to hold a Load Balancing policy for. More details can be found here:

**Comment by daschott** [comment]
Every k8s service created on the cluster + every pod that has outboundNAT policy hosted on the node will reserve contigous range of 64 ports. Which is the case by default for all pods. If pods do not need to reach outside of their own subnet then the outboundNAT section can be removed from azure CNI config. But please only do so if you understand that this means as a consequence pods will not have Internet or DNS access (which OP reported as the original problem)...

On a typical node, it is easier and preferred to change the default available port range using 
`netsh int ipv4 set dynamicportrange TCP <start_port> <port_count>`

**Comment by asbjorne** [comment]
Thanks for the explanations, guys. We've temporarily increased the port range and will also run more clusters to reduce the number of services on each cluster

**Comment by daschott** [comment]
@zhiweiv Depending on your scale needs, there is also DSR load balancing mode shipping later this year to Windows Server 2019 (tentative ETA June 2020) which reserves 0 ports per k8s service and should remove this scale limitation. Internal testing has shown ability to scale to thousands of services that way.

**Comment by zhiweiv** [comment]
That is great, we'd like to test it if there is internal hotfix available soon.

**Comment by MikkelHegn** [comment]
We've increased the range of dynamic ports available in the vhd supporting Windows nodes in AKS. This should set the limit to approx. 500 services.

https://github.com/Azure/AKS/releases/tag/2020-02-10

Closing this issue.

**Comment by zhiweiv** [comment]
@daschott 
Do you have ETA about DSR on Windows Server 2019?
The port exhaustion occurred a few times since Mar 2020 although the workaround was applied to AKS already.

**Comment by daschott** [comment]
@zhiweiv unfortunately the ETA for an update release has slipped out to September due to COVID-19 situation and extensive validation that is needed. That being said, there is a private, test-signed binary that can be shared already for validation/testing should you request it.

**Comment by zhiweiv** [comment]
@daschott 
Will all these paths work in DSR on Windows Server 2019? I saw DSR mentioned in https://techcommunity.microsoft.com/t5/containers/windows-server-version-2004-now-available/ba-p/1419194.

> o container to container traffic
> o container to service traffic
> o container to external traffic (Internet)
> o Host -> service traffic
> o Host -> pod traffic
> o Host -> NodePort traffic
> o External client -> NodePort traffic

**Comment by daschott** [comment]
yes, assuming they worked on non-DSR for the host they should work for DSR as well after the patch has been applied. There will also be a patch released for 1903 to light up these paths that were missing for partial DSR implementation on 1903.
